src/main/scala/scalaz/stream/nondeterminism/nondeterminism.scala

```diff
@@ -113,10 +113,11 @@ package object nondeterminism {
 
             //runs the process with a chance to interrupt it using signal `done`
             //interrupt is done via killing the done signal.
+            //enqueue action needs to be interruptible, otherwise the process will never halt
+            //if the outer process is terminated while the queue is full and an enqueue is pending
             //note that here we convert `Kill` to exception to distinguish form normal and
             //killed behaviour of upstream termination
-            done.discrete.wye(p)(wye.interrupt)
-            .to(q.enqueue)
+            done.discrete.wye(p.to(q.enqueue))(wye.interrupt)
```

Comment:
@edwinchoi excellent catch, thanks. 

I would perhaps for performance reasons to change to 
`p.flatMap(eval_(q.enqueueOne(_))`
That will in fact cause wye to be not consulted for every enqueue, and still shall correctly interrupt the process on done signal. Also will remove the overhead of `tee` that is used by `to` combinator.   

Also please consider to add 
`.onComplete(eval_(q.close))` cleanup code which I think is missing in the current implementation as well.